crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+///
+/// Note that you can also use `Rooted<StructRef>` and
+/// `ManuallyRooted<StructRef>` as a type parameter with
+/// [`Func::typed`][crate::Func::typed]- and
+/// [`Func::wrap`][crate::Func::wrap]-style APIs.
+///
+/// # Example
+///
+/// ```
+/// use wasmtime::*;
+///
+/// # fn foo() -> Result<()> {
+/// let mut config = Config::new();
+/// config.wasm_function_references(true);
+/// config.wasm_gc(true);
+///
+/// let engine = Engine::new(&config)?;
+/// let mut store = Store::new(&engine, ());
+///
+/// // Define a struct type.
+/// let struct_ty = StructType::new(
+///    store.engine(),
+///    [FieldType::new(Mutability::Var, StorageType::I8)],
+/// )?;
+///
+/// // Create an allocator for the struct type.
+/// let allocator = StructRefPre::new(&mut store, struct_ty);
+///
+/// {
+///     let mut scope = RootScope::new(&mut store);
+///
+///     // Allocate an instance of the struct type.
+///     let my_struct = match StructRef::new(&mut scope, &allocator, &[Val::I32(42)]) {
+///         Ok(s) => s,
+///         // If the heap is out of memory, then do a GC and try again.
+///         Err(e) if e.is::<GcHeapOutOfMemory<&'static str>>() => {
+///             // Do a GC! Note: in an async context, you'd want to do
+///             // `scope.as_context_mut().gc_async().await`.
+///             scope.as_context_mut().gc();
+///
+///             StructRef::new(&mut scope, &allocator, &[Val::I32(42)])?
+///         }
+///         Err(e) => return Err(e),
+///     };
+///
+///     // That instance's field should have the expected value.
+///     let val = my_struct.field(&mut scope, 0)?.unwrap_i32();
+///     assert_eq!(val, 42);
+///
+///     // And we can update the field's value because it is a mutable field.
+///     my_struct.set_field(&mut scope, 0, Val::I32(36))?;
+///     let new_val = my_struct.field(&mut scope, 0)?.unwrap_i32();
+///     assert_eq!(new_val, 36);
+/// }
+/// # Ok(())
+/// # }
+/// # foo().unwrap();
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct StructRef {
+    pub(super) inner: GcRootIndex,
+}
+
+unsafe impl GcRefImpl for StructRef {
+    #[allow(private_interfaces)]
+    fn transmute_ref(index: &GcRootIndex) -> &Self {
+        // Safety: `StructRef` is a newtype of a `GcRootIndex`.
+        let me: &Self = unsafe { mem::transmute(index) };
+
+        // Assert we really are just a newtype of a `GcRootIndex`.
+        assert!(matches!(
+            me,
+            Self {
+                inner: GcRootIndex { .. },
+            }
+        ));
+
+        me
+    }
+}
+
+impl StructRef {
+    /// Allocate a new `struct` and get a reference to it.
+    ///
+    /// # Errors
+    ///
+    /// If the allocation cannot be satisfied because the GC heap is currently
+    /// out of memory, but performing a garbage collection might free up space
+    /// such that retrying the allocation afterwards might succeed, then a
+    /// [`GcHeapOutOfMemory<()>`][crate::GcHeapOutOfMemory] error is returned.
```

Comment:
Another source of errors is if `fields` have the wrong types for `allocator`, right?